metricbeat/module/linux/perf/perf.go

```diff
@@ -0,0 +1,277 @@
+	mb.Registry.MustAddMetricSet("linux", "perf", New)
+}
+
+type eventsConfig struct {
+	HardwareEvents bool `config:"hardware"`
+	SoftwareEvents bool `config:"software"`
+}
+
+type sampleConfig struct {
+	ProcessGlob string       `config:"process"`
+	Events      eventsConfig `config:"events"`
+}
+
+// Config holds the metricset config info for perf
+type Config struct {
+	SamplePeriod time.Duration  `config:"perf.sample_period" validate:"required"`
+	Processes    []sampleConfig `config:"perf.processes"`
+}
+
+// perfInfo is the "final" sampling data returned from a sample run
+// The struct contains metadata from the process along with sampling counters
+type perfInfo struct {
+	Metadata  common.MapStr
+	HwMetrics perf.HardwareProfile
+	SwMetrics perf.SoftwareProfile
+}
+
+// procInfo contains all the controlling information on a given PID.
+type procInfo struct {
+	PID          int
+	Metadata     common.MapStr
+	SoftwareProc perf.SoftwareProfiler
+	HardwareProc perf.HardwareProfiler
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	processes  []procInfo
+	period     time.Duration
+	continuous bool
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The linux perf metricset is beta.")
+	linuxModule, ok := base.Module().(*linux.Module)
+	if !ok {
+		return nil, errors.New("unexpected module type")
+	}
+	config := Config{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	if config.SamplePeriod > linuxModule.Period {
+		return nil, fmt.Errorf("Sample period of %s exceeds metricset period of %s", config.SamplePeriod, linuxModule.Period)
+	}
+
+	if config.SamplePeriod == 0 {
+		return nil, fmt.Errorf("Sample Period is zero")
+	}
+
+	var isContinuoius = false
+	if config.SamplePeriod == linuxModule.Period {
+		isContinuoius = true
+	}
+
+	procList, err := matchProcesses(config.Processes)
+	if err != nil {
+		return nil, errors.Wrap(err, "error gathering processes")
+	}
+
+	if len(procList) == 0 {
+		return nil, fmt.Errorf("no processes found that match provided config")
+	}
+
+	if isContinuoius {
+		err := startMonitor(procList)
+		if err != nil {
+			return nil, errors.Wrap(err, "error starting monitor")
+		}
+	}
+
+	// This perf library is Not So Good and eats errors that come from perf_event_open
+	// Do a quck test using a lower-level API to make sure we can actually access perf APIs
+	testP, err := perf.NewMinorFaultsProfiler(procList[0].PID, -1)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating profiler for Minor Page faults")
+	}
+	testP.Close()
```

Comment:
Does this need to be `defer testP.Close()` and moved up so it's right after `testP` is initialized?